components/experiment-creation/Metrics.tsx

```diff
@@ -0,0 +1,290 @@
+  TableRow,
+  TableCell,
+  TableBody,
+  Select as MuiSelect,
+  FormControl,
+  InputLabel,
+  MenuItem,
+  Button,
+  InputAdornment,
+  Tooltip,
+} from '@material-ui/core'
+import { createStyles, makeStyles, Theme } from '@material-ui/core/styles'
+import React, { useState } from 'react'
+import { FieldArray, useField, Field } from 'formik'
+import { MetricAssignment, MetricBare, MetricParameterType, AttributionWindowSeconds } from '@/lib/schemas'
+import { TextField, Select, Switch } from 'formik-material-ui'
+
+import { AttributionWindowSecondsToHuman } from '@/lib/metric-assignments'
+import MoreMenu from '@/components/MoreMenu'
+import { Add } from '@material-ui/icons'
+
+const normalizedMetrics: Record<number, MetricBare> = {
+  1: {
+    metricId: 1,
+    name: 'asdf_7d_refund',
+    description: 'string',
+    parameterType: MetricParameterType.Revenue,
+  },
+  2: {
+    metricId: 2,
+    name: 'registration_start',
+    description: 'string',
+    parameterType: MetricParameterType.Conversion,
+  },
+}
+
+const useStyles = makeStyles((theme: Theme) =>
+  createStyles({
+    root: {},
+    addMetric: {
+      display: 'flex',
+      justifyContent: 'center',
+      alignItems: 'flex-end',
+      margin: theme.spacing(3, 0, 2),
+    },
+    addMetricAddSymbol: {
+      position: 'relative',
+      top: -3,
+      marginRight: theme.spacing(2),
+      color: theme.palette.text.disabled,
+    },
+    addMetricPlaceholder: {
+      fontFamily: theme.typography.fontFamily,
+    },
+    addMetricSelect: {
+      minWidth: '10rem',
+      marginRight: theme.spacing(1),
+    },
+    attributionWindowSelect: {
+      minWidth: '8rem',
+    },
+    metricName: {
+      fontFamily: theme.custom.fonts.monospace,
+      fontWeight: theme.custom.fontWeights.monospaceBold,
+    },
+    primary: {
+      fontFamily: theme.custom.fonts.monospace,
+      opacity: 0.5,
+    },
+    minDifferenceField: {
+      maxWidth: '14rem',
+    },
+    changeExpected: {
+      textAlign: 'center',
+    },
+  }),
+)
+
+const createMetricAssignment = (metric: MetricBare) => {
+  return {
+    metricId: metric.metricId,
+    attributionWindowSeconds: '',
+    isPrimary: false,
+    changeExpected: false,
+    minDifference: '',
+  }
+}
+
+const Metrics = () => {
+  const classes = useStyles()
+
+  const [metricAssignmentsField, _metricAssignmentsFieldMetaProps, metricAssignmentsFieldHelperProps] = useField<
+    MetricAssignment[]
+  >('experiment.metricAssignments')
+  const [selectedMetricId, setSelectedMetricId] = useState<string>('')
+  const onSelectedMetricChange = (event: React.ChangeEvent<{ value: unknown }>) => {
+    setSelectedMetricId(event.target.value as string)
+  }
+
+  const makeMetricAssignmentPrimary = (indexToSet: number) => {
+    metricAssignmentsFieldHelperProps.setValue(
+      metricAssignmentsField.value.map((metricAssignment, index) => ({
+        ...metricAssignment,
+        isPrimary: index === indexToSet,
+      })),
+    )
+  }
```

Comment:
Would it be better to pass the `metricAssignment` instead of the index?

```suggestion
  const makeMetricAssignmentPrimary = (metricToSet) => {
    metricAssignmentsFieldHelperProps.setValue(
      metricAssignmentsField.value.map((metricAssignment) => ({
        ...metricAssignment,
        isPrimary: metricAssignment === metricToSet,
      })),
    )
  }
```

Reply:
Ah interesting, using the object itself as the "id".

I don't see any particular advantages to doing so and it seems to me a bit less direct and less idiomatic.

And for a disadvantage: What if someone in the future immutably changes a metricAssignment before calling `makeMetricAssignmentPrimary`

More generally, in JS I would prefer to work only at the "value" level.
- Using an object like this assigns an identity to a value that is seperate from the value, the identity should be within the value so that values are comparable. Now you would have two kinds of identity - object and value, why not have one.
- It is hidden so it can't be saved, pushed to the server, or logged.

Exceptions to this might be for performance sensitive code.